File: day2_part2_2019.R

program <- scan("input.txt", what = integer(), sep = ",", quiet = TRUE)

intcode_computer <- function(program, noun, verb) {
  program[2] <- noun
  program[3] <- verb
  
  i <- 1
  while (program[i] != 99) {
    if (program[i] == 1) {
      program[program[i + 3] + 1] <- program[program[i + 1] + 1] + program[program[i + 2] + 1]
    } else if (program[i] == 2) {
      program[program[i + 3] + 1] <- program[program[i + 1] + 1] * program[program[i + 2] + 1]
    }
    i <- i + 4
  }
  
  return(program[1])
}

# Part One
result_part_one <- intcode_computer(program, 12, 2)
print(result_part_one)

# Part Two
target_output <- 19690720
for (noun in 0:99) {
  for (verb in 0:99) {
    if (intcode_computer(program, noun, verb) == target_output) {
      result_part_two <- 100 * noun + verb
      print(result_part_two)
      break
    }
  }
}

File: day2_part2_2019.pas
program GravityAssist;

{$mode objfpc}{$H+}

uses
  SysUtils, Classes;

var
  inputFile: TextFile;
  originalIntcode, intcode: array of Integer;
  noun, verb, output: Integer;
  line: string;
  numbers: TStringArray;
  i: Integer;

function ExecuteIntcode(memory: array of Integer): Integer;
var
  position: Integer;
begin
  position := 0;
  while (position < Length(memory)) and (memory[position] <> 99) do
  begin
    case memory[position] of
      1: memory[memory[position + 3]] := memory[memory[position + 1]] + memory[memory[position + 2]];
      2: memory[memory[position + 3]] := memory[memory[position + 1]] * memory[memory[position + 2]];
    end;
    Inc(position, 4);
  end;
  Result := memory[0];
end;

begin
  // Load the Intcode program from the input file
  AssignFile(inputFile, 'input.txt');
  Reset(inputFile);
  ReadLn(inputFile, line);
  CloseFile(inputFile);

  numbers := line.Split([',']);
  SetLength(originalIntcode, Length(numbers));
  for i := 0 to High(numbers) do
  begin
    originalIntcode[i] := StrToInt(numbers[i]);
  end;

  output := 0;

  // Try every combination of noun and verb
  for noun := 0 to 99 do
    for verb := 0 to 99 do
    begin
      // Reset memory to the original program, with the current noun and verb
      intcode := Copy(originalIntcode, 0, Length(originalIntcode));
      intcode[1] := noun;
      intcode[2] := verb;

      // Run the program
      if ExecuteIntcode(intcode) = 19690720 then
      begin
        output := 100 * noun + verb;
        Break;
      end;
    end;

  if output > 0 then
    WriteLn('The value that produces the output 19690720 is: ', output)
  else
    WriteLn('No combination of noun and verb found that produces the output 19690720.');
end.
